File: java/iroha_android/src/main/java/org/hyperledger/iroha/android/model/instructions/RegisterVerifyingKeyInstruction.java
package org.hyperledger.iroha.android.model.instructions;

import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Objects;
import org.hyperledger.iroha.android.model.InstructionBox;
import org.hyperledger.iroha.android.model.zk.VerifyingKeyRecordDescription;

/** Typed builder for {@code RegisterVerifyingKey} instructions. */
public final class RegisterVerifyingKeyInstruction implements InstructionTemplate {

  private static final String ACTION = "RegisterVerifyingKey";

  private final String backend;
  private final String name;
  private final VerifyingKeyRecordDescription record;
  private final Map<String, String> arguments;

  private RegisterVerifyingKeyInstruction(final Builder builder, final Map<String, String> args) {
    this.backend = builder.backend;
    this.name = builder.name;
    this.record = builder.record;
    this.arguments = Map.copyOf(args);
  }

  public String backend() {
    return backend;
  }

  public String name() {
    return name;
  }

  public VerifyingKeyRecordDescription record() {
    return record;
  }

  @Override
  public InstructionKind kind() {
    return InstructionKind.CUSTOM;
  }

  @Override
  public Map<String, String> toArguments() {
    return arguments;
  }

  public static RegisterVerifyingKeyInstruction fromArguments(final Map<String, String> arguments) {
    final String backend = VerifyingKeyInstructionUtils.require(arguments, "backend");
    final String name = VerifyingKeyInstructionUtils.require(arguments, "name");
    final VerifyingKeyRecordDescription record =
        VerifyingKeyInstructionUtils.parseRecord(arguments, backend);
    final Builder builder =
        builder().setBackend(backend).setName(name).setRecord(record);
    return builder.build();
  }

  public static Builder builder() {
    return new Builder();
  }

  @Override
  public boolean equals(final Object obj) {
    if (this == obj) {
      return true;
    }
    if (!(obj instanceof RegisterVerifyingKeyInstruction other)) {
      return false;
    }
    return Objects.equals(backend, other.backend)
        && Objects.equals(name, other.name)
        && Objects.equals(record, other.record);
  }

  @Override
  public int hashCode() {
    return Objects.hash(backend, name, record);
  }

  public static final class Builder {
    private String backend;
    private String name;
    private VerifyingKeyRecordDescription record;

    private Builder() {}

    public Builder setBackend(final String backend) {
      if (backend == null || backend.trim().isEmpty()) {
        throw new IllegalArgumentException("backend must not be blank");
      }
      this.backend = backend.trim();
      return this;
    }

    public Builder setName(final String name) {
      if (name == null || name.trim().isEmpty()) {
        throw new IllegalArgumentException("name must not be blank");
      }
      this.name = name.trim();
      return this;
    }

    public Builder setRecord(final VerifyingKeyRecordDescription record) {
      this.record = Objects.requireNonNull(record, "record");
      return this;
    }

    public RegisterVerifyingKeyInstruction build() {
      if (backend == null) {
        throw new IllegalStateException("backend must be provided");
      }
      if (name == null) {
        throw new IllegalStateException("name must be provided");
      }
      if (record == null) {
        throw new IllegalStateException("record must be provided");
      }
      final Map<String, String> args = new LinkedHashMap<>();
      args.put("action", ACTION);
      args.put("backend", backend);
      args.put("name", name);
      args.putAll(record.toArguments(backend));
      return new RegisterVerifyingKeyInstruction(this, args);
    }
  }
}

File: kotlin/core-jvm/src/main/java/org/hyperledger/iroha/sdk/core/model/instructions/RegisterVerifyingKeyInstruction.kt
package org.hyperledger.iroha.sdk.core.model.instructions

import org.hyperledger.iroha.sdk.core.model.instructions.VerifyingKeyInstructionUtils.nonEmptyString
import org.hyperledger.iroha.sdk.core.model.instructions.VerifyingKeyInstructionUtils.parseRecord
import org.hyperledger.iroha.sdk.core.model.zk.VerifyingKeyRecordDescription

private const val ACTION = "RegisterVerifyingKey"

/** Typed representation of `RegisterVerifyingKey` instructions. */
class RegisterVerifyingKeyInstruction private constructor(
    @JvmField val backend: String,
    @JvmField val name: String,
    @JvmField val record: VerifyingKeyRecordDescription,
    override val arguments: Map<String, String>,
) : InstructionTemplate {

    constructor(
        backend: String,
        name: String,
        record: VerifyingKeyRecordDescription,
    ) : this(
        backend = validatedNotBlank(backend, "backend"),
        name = validatedNotBlank(name, "name"),
        record = record,
        arguments = buildArguments(backend.trim(), name.trim(), record),
    )

    override val kind: InstructionKind get() = InstructionKind.CUSTOM

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is RegisterVerifyingKeyInstruction) return false
        return backend == other.backend
            && name == other.name
            && record == other.record
    }

    override fun hashCode(): Int {
        var result = backend.hashCode()
        result = 31 * result + name.hashCode()
        result = 31 * result + record.hashCode()
        return result
    }

    companion object {
        @JvmStatic
        fun builder(): Builder = Builder()

        @JvmStatic
        fun fromArguments(arguments: Map<String, String>): RegisterVerifyingKeyInstruction {
            val backend = arguments.nonEmptyString("backend")
            val name = arguments.nonEmptyString("name")
            val record = arguments.parseRecord(backend)
            return RegisterVerifyingKeyInstruction(backend, name, record)
        }

        private fun buildArguments(
            backend: String,
            name: String,
            record: VerifyingKeyRecordDescription,
        ): Map<String, String> {
            val args = linkedMapOf<String, String>()
            args["action"] = ACTION
            args["backend"] = backend
            args["name"] = name
            args.putAll(record.toArguments(backend))
            return args
        }

        private fun validatedNotBlank(value: String, name: String): String {
            val trimmed = value.trim()
            require(trimmed.isNotEmpty()) { "$name must not be blank" }
            return trimmed
        }
    }

    class Builder internal constructor() {
        private var backend: String? = null
        private var name: String? = null
        private var record: VerifyingKeyRecordDescription? = null

        fun setBackend(backend: String) = apply {
            require(backend.trim().isNotEmpty()) { "backend must not be blank" }
            this.backend = backend.trim()
        }

        fun setName(name: String) = apply {
            require(name.trim().isNotEmpty()) { "name must not be blank" }
            this.name = name.trim()
        }

        fun setRecord(record: VerifyingKeyRecordDescription) = apply {
            this.record = requireNotNull(record) { "record" }
        }

        fun build(): RegisterVerifyingKeyInstruction {
            val b = checkNotNull(backend) { "backend must be provided" }
            val n = checkNotNull(name) { "name must be provided" }
            val r = checkNotNull(record) { "record must be provided" }
            return RegisterVerifyingKeyInstruction(b, n, r)
        }
    }
}
